ngraph/core/reference/include/ngraph/runtime/reference/adaptive_max_pool.hpp

```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cfenv>
+#include <cmath>
+#include <numeric>
+#include <stdexcept>
+#include <vector>
+
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            template <typename T>
+            void adaptive_max_pool_1d(
+                const T* arg, T* out, int64_t* indices, size_t h_in, size_t h_out)
+            {
+                for (size_t i = 0; i < h_out; i++)
+                {
+                    auto from = arg + adaptive_pool::window_start(i, h_in, h_out);
+                    auto to = arg + adaptive_pool::window_end(i, h_in, h_out);
+                    NGRAPH_CHECK(to - from != 0, "AdaptiveMaxPool elements == 0, must be non-zero");
+                    auto it = std::max_element(from, to);
+                    out[i] = *it;
+                    indices[i] = it - arg;
+                }
+            }
+            template <typename T>
+            void adaptive_max_pool_2d(const T* arg,
+                                      T* out,
+                                      int64_t* indices,
+                                      size_t h_in,
+                                      size_t h_out,
+                                      size_t w_in,
+                                      size_t w_out)
+            {
+                for (size_t i = 0; i < h_out; i++)
+                {
+                    size_t h_start = adaptive_pool::window_start(i, h_in, h_out);
+                    size_t h_end = adaptive_pool::window_end(i, h_in, h_out);
+                    for (size_t j = 0; j < w_out; j++)
+                    {
+                        size_t w_start = adaptive_pool::window_start(j, w_in, w_out);
+                        size_t w_end = adaptive_pool::window_end(j, w_in, w_out);
+                        NGRAPH_CHECK((w_end - w_start) * (h_end - h_start) != 0,
+                                     "AdaptiveMaxPool elements == 0, must be non-zero");
+                        auto result = arg + h_start * w_in + w_start;
+                        for (size_t n = h_start; n < h_end; n++)
+                        {
+                            auto from = arg + n * w_in + w_start;
+                            auto to = arg + n * w_in + w_end;
+                            auto it = std::max_element(from, to);
+                            result = *it > *result ? it : result;
```

Comment:
It looks like `std::max`.

Reply:
it is different... Cant replace with std::max

Reply:
Now I see.